### ISSUE TITLE
Migrate to Redux Persist v5

### ISSUE THREAD

**Issue by MrLoh** [redux-offline/redux-offline#119]
Are there any plans to migrate to redux-persist v5?

**Comment by ScreamZ** [comment]
Any news about this ?

Was reading about that and I'm a bit afraid on how to deal with that https://shift.infinite.red/shipping-persistant-reducers-7341691232b1 ? Using redux-offline at the moment ?

@MrLoh Are you using it ?

**Comment by wacii** [comment]
If you don't want Redux Offline to setup Redux Persist, provide a falsey value for `offlineConfig.persist`. You are then free to handle setup however you like.

And Redux Persist v5 is still only a release candidate.

**Comment by rt2zz** [comment]
does redux-offline store its working state in its own reducer? redux-persist v5 has a `persistReducer` higher order reducer, so depending on how redux-offline manages its internal state will affect I think how the integration will look.

**Comment by wacii** [comment]
Redux Offline uses a higher-order reducer to store its working state under `state.offline`. `state.offline.outbox` needs to be persisted for retries to work across sessions.

**Comment by rt2zz** [comment]
ok sounds like integration should be pretty straightforward. redux persist V5 is just about ready, I think if we can get a PR up to upgrade redux-offline with no API changes required then I will be confident to release V5.0.0. I might take a look this weekend, buy by all means if someone wants to beat me to it please do.

**Comment by fungilation** [comment]
Awesome. And good to see that `redux-offline` isn't dead with inactivity, since there's an active fork. Thanks to you both @wacii @rt2zz for your work!

I've been watching your repos, as I plan on adding offline persistence to my React Native app. `redux-persist` v5 looks like a breaking change that I'm waiting for stabilizing first.

**Comment by rt2zz** [comment]
FYI I have not looked at this yet, but I am pretty confident upgrading to v5 will be straightforward process of adding persistReducer around the reducer redux-offline creates, with `whitelist: ['offline']` in the config.

**Comment by wacii** [comment]
I played around with this and it seemed to come together without issue. Code [here](https://codesandbox.io/s/vq429o84w7).

@sorodrigo Not sure what to do with this. I'm leaning to adding an example and linking to it in the README. Its a little too much code to dump in the README directly, although we could start breaking off chunks into separate wiki pages.

And I don't think we should migrate Redux Offline itself. It would seem the new version of Redux Persist doesn't work with a top-level immutable object or HMR, which are both things we do here. And since v4 is still going to be supported, I figure we can stay with that for now.

**Comment by fungilation** [comment]
https://github.com/rt2zz/redux-persist#why-v5

> * easier to ship persist support inside of other libs (**e.g. redux-offline**)

That alone isn't reason enough? I for one care about PersistGate and migration support.

If moving to v5 is "easy", the standard thing to deal with downstream breaking change is move `redux-offline` to a major new version (3.x) that coincides with the move to v5, and keep `redux-offline` 2.x on v4.

**Comment by sorodrigo** [comment]
@fungilation the issue is not that there’s a downstream breaking change as you put it; we would lose features we do want to support. So as @wacii states, the best thing is to make the v5 upgrade an opt-in feature.

@wacii about the readme, I agree we should start doing something else. A wiki is agood idea. Maybe we could copy from another lib that manages this well? Lets open another issue where we can discuss this more thoroughly.

**Comment by MrLoh** [comment]
Redux-offline is even mentioned in the redux-persist v5 docs under ‘Why v5’

> easier to ship persist support inside of other libs (e.g. redux-offline)

So there must be a good reason for redux-offline to upgrade, or the redux persist guys don’t understand your needs.

**Comment by gusgard** [comment]
@sorodrigo @wacii Thanks. Do you know how can I sync `REHYDRATE` to be executed after`Offline/STATUS_CHANGED`. `persistStore` expect a callback as a third param but don't know how can I get it.

Before 
<img width="203" alt="screen shot 2018-02-09 at 9 03 26 pm" src="https://user-images.githubusercontent.com/2577356/36055676-bfcee06c-0ddc-11e8-8901-e556d4c9afec.png">

Now (with version 5)

<img width="184" alt="screen shot 2018-02-09 at 8 58 00 pm" src="https://user-images.githubusercontent.com/2577356/36055677-c1df0d28-0ddc-11e8-8e13-d47ef8d86f89.png">


Solution: 
```
const offlineConfig = {
  ...defaultConfig,
  persist: store => {
    const persistor = persistStore(store);
    ...
  }
};
```

Is the migration to v5 in the roadmap?

**Comment by wacii** [comment]
@gusgard There is no built in way to do this. You will have to initialize the network detector yourself.

This is how it's done now: [index.js](https://github.com/redux-offline/redux-offline/blob/develop/src/index.js#L70-L75). You will want to pass a falsey value for `config.detectNetwork`, then set it up yourself in the persist callback.

As for migration to v5, unless something has changed, no. I talked about this a bit in my previous comment.

**Comment by tadeo** [comment]
Hi all, any news about this? Thanks!

**Comment by antonsivogrivov** [comment]
Any updates?

**Comment by wacii** [comment]
@antonsivogrivov No. I explained why we don't intend to migrate at the moment above. If something has changed with Redux Persist I would be happy to take another look at this. Otherwise now that the documentation has been updated I think an article explaining how to integrate with Redux Persist v5 would be an excellent addition. Not sure when I would get to this myself, but PRs are always welcome.

**Comment by jarvisluong** [comment]
@wacii I just created a PR to propose a guide for helping with config redux-persist v5 😄

**Comment by jfbloom22** [comment]
@jarvisluong thanks for the guide to use redux-persist v5!  I would really like to use the new version for the migration support.  However, I am stuck on where you are returning the persister and the store:
`return {persistor, store};` on line 42. I don't understand why this does not return a single `store`? 
When I try to pass the result of `configureStore()` to my provider, like this: `<Provider store={store}>`,  typescript is giving me this error:  
```Type '{ persistor: Persistor; store: Store<any, any>; }' is not assignable to type 'Store<any, AnyAction>'```


Below is the code where I am trying to follow your example:  
```
const persistConfig = {
    key: 'my-key',
    debounce: 500,
    storage: localForage
}

const {
  middleware: offlineMiddleware,
  enhanceReducer: offlineEnhanceReducer,
  enhanceStore: offlineEnhanceStore
} = createOffline({
  ...offlineConfig,
  persist: false,
  effect,
  discard

});


const persistedReducer = persistReducer(
  persistConfig,
  offlineEnhanceReducer(rootReducer)
);

export default function configureStore() {
    const store = createStore(
      persistedReducer,
      require('redux-devtools-extension').composeWithDevTools(
        offlineEnhanceStore,
        applyMiddleware(
          thunk as ThunkMiddleware<IinitialState, any>,
          require('redux-immutable-state-invariant').default(),
          offlineMiddleware as Middleware<any, any, any>
        )
      )
    ) ;
    const persistor = persistStore(store);
    return {persistor, store};
```

I am also not sure if I am passing in my custom `effect` and `discard` callbacks properly.

**Comment by geminiyellow** [comment]
this issue last too long. hope redux-offline can upgrade it.

**Comment by jfbloom22** [comment]
I figured out why @jarvisluong was returning `return {persistor, store};`.  If you take a look at the redux-persist documentation they describe how to use the `persistor` when using React:  https://github.com/rt2zz/redux-persist#basic-usage

Here is the important snippets that I added to my index.tsx:  
```
import { PersistGate } from 'redux-persist/integration/react'
const {store, persistor} = configureStore();
const App = () => {
  return (
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <RootComponent />
      </PersistGate>
    </Provider>
  );
};
```

Now everything seems to be working great!
Here is a gist of my configureStore.ts:  https://gist.github.com/jfbloom22/6ef92d9ff28007be8ce1154d28e7c66e

**Comment by meodemsao** [comment]
This is my config 

```javascript
import { createStore, applyMiddleware, compose, combineReducers } from 'redux'
import createHistory from 'history/createBrowserHistory'
import { routerMiddleware } from 'react-router-redux'
import thunk from 'redux-thunk'
import { persistStore, persistReducer } from 'redux-persist'
import immutableTransform from 'redux-persist-transform-immutable'
import storage from 'redux-persist/lib/storage'
import createGraphQLSubscriptionsMiddleware from './vendor/redux-graphql-subscriptions'
import { createOffline } from '@redux-offline/redux-offline';
import offlineConfig from '@redux-offline/redux-offline/lib/defaults/index';

import reducers from './reducers' // Or wherever you keep your reducers

// create graphql subscription middleware
const graphQLSubscriptionsMiddleware = createGraphQLSubscriptionsMiddleware(process.env.REACT_APP_WS_URL || 'ws://localhost:4000')

const persistConfig = {
  key: 'root',
  storage: storage,
  transforms: [immutableTransform()],
  blacklist: ['data', 'notification']
}

const {
  middleware: offlineMiddleware,
  enhanceReducer: offlineEnhanceReducer,
  enhanceStore: offlineEnhanceStore
} = createOffline({
  ...offlineConfig,
  persist: false
});

// persisted reducer
const persistedReducer = persistReducer(persistConfig, offlineEnhanceReducer(combineReducers(reducers)))

// Create a history of your choosing (we're using a browser history in this case)
const history = createHistory()

// Build the middleware for intercepting and dispatching navigation actions
// thunk
// offline middleware
// router middleware
// graphql subscriptions middleware
const middleware = [thunk, offlineMiddleware, routerMiddleware(history), graphQLSubscriptionsMiddleware];

// create store
// persist reducer
// offline store
// middleware
const store = createStore(persistedReducer, {}, compose(offlineEnhanceStore, applyMiddleware(...middleware)))

const persistor = persistStore(store)

export default { history, store, persistor }
`

**Comment by pdpino** [comment]
There is a new reason to migrate to redux-persist v5 (and to redux-persist v6, which is ~~soon to come~~ already released, see below).

React native v0.59.0 (released March 12, 2019) is deprecating `AsyncStorage` as is being moved to `@react-native-community/asyncstorage` (see [release notes](https://github.com/react-native-community/releases/blob/master/CHANGELOG.md#059) and [standalone repo](https://github.com/react-native-community/react-native-async-storage)).

When using RN >= 0.59 and redux-offline (2.5.1), you get a warning indicating the deprecation:
```
Warning: Async Storage has been extracted from react-native core and will be removed in a
future release. It can now be installed and imported
from '@react-native-community/async-storage' instead of 'react-native'.
See https://github.com/react-native-community/react-native-async-storage
```
See [this issue](https://github.com/rt2zz/redux-persist/issues/1013) in redux-persist, which also shows the warning.

There is a redux-persist [pre-release](https://github.com/rt2zz/redux-persist/releases/tag/v6.0.0-pre1) (v6.0.0) addressing this change. To be compatible with redux-persist v6, compatibility with v5 is needed.

UPDATE: redux-persist v6.0.0 was [released](https://github.com/rt2zz/redux-persist/releases/tag/v6.0.0), as of september 1th 2019.

**Comment by calumpeak** [comment]
@pdpino Thanks for the update. Will take a look at getting this sorted soon.

**Comment by DianaLaa** [comment]
There are a few issues that can be encountered when using v5, including #239 and #151. For me, these disappeared when reverting to redux-persist v4 that is included in redux-offline. Also, I'd very much like to make use of new redux-persist functionality, such as nested whitelisting/blacklisting. 

Is there a plan of attack for this upgrade? Some schedule? I fully understand that time spent on developing redux-offline may be limited, but some ballpark guess as to when v5 / v6 can be integrated would be very helpful.

**Comment by calumpeak** [comment]
@DianaKoenraadt Hi 👋 

We have V3 in the pipeline for redux-offline, unfortunately there's been no movement due to all the main contributors (including myself) having busy schedules/other jobs. Hopefully we can get a start on this again soon. We don't have any timelines to provide at the moment I'm afraid.

**Comment by sorodrigo** [comment]
I want to come back and spend some time on this, could we get a new thread going with what we would like to have for a v3?
- typescript migration(?)
- redux-persist new versions
- decouple lib from react and react-native
- background sync(? is this even possible?)
- multiple tabs issue(? is this even possible?)

**Comment by calumpeak** [comment]
@sorodrigo I need to log back into our slack again and we can get this going. But yes, all of those would be really nice to address - if we can make the system more pluggable, we can provide modules which can encapsulate a lot of that logic (and allow quick upgrading of things like redux-persist) and allow users to write their own if necessary. E.g. I think multiple tabs and background sync can be solved with service-workers, but this differs completely from Native where you'd be able to plug another module in for that)

**Comment by sorodrigo** [comment]
@DianaKoenraadt would you like to get involved in the development of v3?

**Comment by DianaLaa** [comment]
@sorodrigo no

**Comment by ankeetmaini** [comment]
I'd love to be a part of this!

**Comment by jfbloom22** [comment]
TLDR; Give redux-persist v4 a try before using v5 or v6.  After painstakingly getting v5 then v6 to work I ended up reverting back to v4 and could not be happier.  

I have a large redux store and was running into performance issues persisting that to a single indexDB key.  I tried nested persists, but I was also not able to get migrations working.

 I ended up reverting back to redux-persist v4 which automatically stores each root reducer in a separate key in indexDB.  The main challenge was reworking my configureStore() to use redux-persist-migrate.  It kinda feel bad about how pleased I am with redux-persist v4.   My configurations are simpler and the performance is significantly better.  The initial load of my PWA offline capable app loads around 75GB of JSON after initial login.  The time it takes for this went from 34 seconds to 13 seconds.  Whoop!

**Comment by sorodrigo** [comment]
That’s interesting. Thanks for sharing. I’m starting to work on v3. And I think we will be compatible with all versions of redux persist. 😀
